python/ql/src/semmle/python/Pruning.qll

```diff
@@ -551,9 +547,23 @@ module Pruner {
     /* Helper for `unreachableEdge`, deal with inequalities here to avoid blow up */
     pragma [inline]
     private predicate contradicts(Constraint a, Constraint b) {
-        a.contradicts(b) or
-        a.(ConstrainedByConstant).minValue() > b.(ConstrainedByConstant).maxValue() or
+        a = TIsNone(true) and b.cannotBeNone()
+        or
+        a.cannotBeNone() and b = TIsNone(true)
+        or
+        a.constrainsVariableToBe(true) and b.constrainsVariableToBe(false)
+        or
+        a.constrainsVariableToBe(false) and b.constrainsVariableToBe(true)
+        or
+        a.(ConstrainedByConstant).minValue() > b.(ConstrainedByConstant).maxValue()
+        or
         a.(ConstrainedByConstant).maxValue() < b.(ConstrainedByConstant).minValue()
+        or
+        exists(int val |
+            a.(ConstrainedByConstant).eq(val) and b.(ConstrainedByConstant).ne(val)
+            or
+            a.(ConstrainedByConstant).ne(val) and b.(ConstrainedByConstant).eq(val)
+        )
```

Comment:
Why not leave out every second disjunct, and add `or contradicts(b, a)` at the end?

Reply:
That would make it recursive.

Reply:
Ah, I didn't know `inline`d predicates couldn't be recursive (though, thinking about it, that does make sense).

In that case, you _could_ do something like this:
```java
    pragma [inline]
    private predicate contradicts(Constraint a, Constraint b) {
        excludes(a, b) or excludes(b, a)
    }

    pragma [inline]
    private predicate excludes(Constraint a, Constraint b) {
        a = TIsNone(true) and b.cannotBeNone()
        or
        a.constrainsVariableToBe(true) and b.constrainsVariableToBe(false)
        or
        a.(ConstrainedByConstant).minValue() > b.(ConstrainedByConstant).maxValue()
        or
        exists(int val |
            a.(ConstrainedByConstant).eq(val) and b.(ConstrainedByConstant).ne(val)
        )
    }
```
But I'm not sure it's much of an improvement. I'll let you decide. If you think it's better as-is, I'll go ahead with the merge.

Reply:
Your version is a bit shorter, but mine more clearly lists all the cases, so I think I'd rather leave it as it is.